{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lab 2, Part 1 - Code & Comments\n",
    "Elysa Strunin  \n",
    "October 2018  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from pandas.io.json import json_normalize\n",
    "\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score, f1_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn import preprocessing\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.preprocessing import label_binarize\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset_version</th>\n",
       "      <th>IC50_Results_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>MAX_CONC_MICROMOLAR</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17</td>\n",
       "      <td>335</td>\n",
       "      <td>924100</td>\n",
       "      <td>1026</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>336</td>\n",
       "      <td>924100</td>\n",
       "      <td>1028</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>337</td>\n",
       "      <td>924100</td>\n",
       "      <td>1029</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.34</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17</td>\n",
       "      <td>338</td>\n",
       "      <td>924100</td>\n",
       "      <td>1030</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.16</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17</td>\n",
       "      <td>339</td>\n",
       "      <td>924100</td>\n",
       "      <td>1031</td>\n",
       "      <td>0.2</td>\n",
       "      <td>-4.33</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Dataset_version  IC50_Results_ID  COSMIC_ID  DRUG_ID  MAX_CONC_MICROMOLAR  \\\n",
       "0               17              335     924100     1026                  1.0   \n",
       "1               17              336     924100     1028                  2.0   \n",
       "2               17              337     924100     1029                  2.0   \n",
       "3               17              338     924100     1030                 10.0   \n",
       "4               17              339     924100     1031                  0.2   \n",
       "\n",
       "   LN_IC50   AUC  RMSE  \n",
       "0     0.72  0.90  0.11  \n",
       "1     2.66  0.96  0.18  \n",
       "2     3.34  0.97  0.08  \n",
       "3     5.16  0.98  0.09  \n",
       "4    -4.33  0.51  0.09  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dose_response = pd.read_csv('v17_fitted_dose_response.csv')\n",
    "dose_response.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_SYMBOLS</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>...</th>\n",
       "      <th>LINC00526</th>\n",
       "      <th>PPY2</th>\n",
       "      <th>Unnamed: 17730</th>\n",
       "      <th>Unnamed: 17731</th>\n",
       "      <th>KRT18P55</th>\n",
       "      <th>Unnamed: 17733</th>\n",
       "      <th>POLRMTP1</th>\n",
       "      <th>UBL5P2</th>\n",
       "      <th>TBC1D3P5</th>\n",
       "      <th>Unnamed: 17737</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENE_title</td>\n",
       "      <td>tetraspanin 6 [Source:HGNC Symbol;Acc:11858]</td>\n",
       "      <td>tenomodulin [Source:HGNC Symbol;Acc:17757]</td>\n",
       "      <td>dolichyl-phosphate mannosyltransferase polypep...</td>\n",
       "      <td>SCY1-like 3 (S. cerevisiae) [Source:HGNC Symbo...</td>\n",
       "      <td>chromosome 1 open reading frame 112 [Source:HG...</td>\n",
       "      <td>Gardner-Rasheed feline sarcoma viral (v-fgr) o...</td>\n",
       "      <td>complement factor H [Source:HGNC Symbol;Acc:4883]</td>\n",
       "      <td>fucosidase, alpha-L- 2, plasma [Source:HGNC Sy...</td>\n",
       "      <td>glutamate-cysteine ligase, catalytic subunit [...</td>\n",
       "      <td>...</td>\n",
       "      <td>long intergenic non-protein coding RNA 526 [So...</td>\n",
       "      <td>pancreatic polypeptide 2 [Source:HGNC Symbol;A...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>keratin 18 pseudogene 55 [Source:HGNC Symbol;A...</td>\n",
       "      <td>hsa-mir-5195 [Source:miRBase;Acc:MI0018174]</td>\n",
       "      <td>polymerase (RNA) mitochondrial (DNA directed) ...</td>\n",
       "      <td>ubiquitin-like 5 pseudogene 2 [Source:HGNC Sym...</td>\n",
       "      <td>TBC1 domain family, member 3 pseudogene 5 [Sou...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DATA.906826</td>\n",
       "      <td>7.63202317146339</td>\n",
       "      <td>2.96458512058924</td>\n",
       "      <td>10.3795526353077</td>\n",
       "      <td>3.61479404843988</td>\n",
       "      <td>3.38068143582194</td>\n",
       "      <td>3.32469189081249</td>\n",
       "      <td>3.56635031211478</td>\n",
       "      <td>8.20453043072163</td>\n",
       "      <td>5.2351175492262</td>\n",
       "      <td>...</td>\n",
       "      <td>6.78692491488045</td>\n",
       "      <td>2.99705393632274</td>\n",
       "      <td>3.109774</td>\n",
       "      <td>7.882377</td>\n",
       "      <td>3.33113425699397</td>\n",
       "      <td>2.85253711034634</td>\n",
       "      <td>3.13069614031282</td>\n",
       "      <td>9.98661583035146</td>\n",
       "      <td>3.07372352861326</td>\n",
       "      <td>7.284733</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 17738 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  GENE_SYMBOLS                                        TSPAN6  \\\n",
       "0   GENE_title  tetraspanin 6 [Source:HGNC Symbol;Acc:11858]   \n",
       "1  DATA.906826                              7.63202317146339   \n",
       "\n",
       "                                         TNMD  \\\n",
       "0  tenomodulin [Source:HGNC Symbol;Acc:17757]   \n",
       "1                            2.96458512058924   \n",
       "\n",
       "                                                DPM1  \\\n",
       "0  dolichyl-phosphate mannosyltransferase polypep...   \n",
       "1                                   10.3795526353077   \n",
       "\n",
       "                                               SCYL3  \\\n",
       "0  SCY1-like 3 (S. cerevisiae) [Source:HGNC Symbo...   \n",
       "1                                   3.61479404843988   \n",
       "\n",
       "                                            C1orf112  \\\n",
       "0  chromosome 1 open reading frame 112 [Source:HG...   \n",
       "1                                   3.38068143582194   \n",
       "\n",
       "                                                 FGR  \\\n",
       "0  Gardner-Rasheed feline sarcoma viral (v-fgr) o...   \n",
       "1                                   3.32469189081249   \n",
       "\n",
       "                                                 CFH  \\\n",
       "0  complement factor H [Source:HGNC Symbol;Acc:4883]   \n",
       "1                                   3.56635031211478   \n",
       "\n",
       "                                               FUCA2  \\\n",
       "0  fucosidase, alpha-L- 2, plasma [Source:HGNC Sy...   \n",
       "1                                   8.20453043072163   \n",
       "\n",
       "                                                GCLC      ...        \\\n",
       "0  glutamate-cysteine ligase, catalytic subunit [...      ...         \n",
       "1                                    5.2351175492262      ...         \n",
       "\n",
       "                                           LINC00526  \\\n",
       "0  long intergenic non-protein coding RNA 526 [So...   \n",
       "1                                   6.78692491488045   \n",
       "\n",
       "                                                PPY2 Unnamed: 17730  \\\n",
       "0  pancreatic polypeptide 2 [Source:HGNC Symbol;A...            NaN   \n",
       "1                                   2.99705393632274       3.109774   \n",
       "\n",
       "  Unnamed: 17731                                           KRT18P55  \\\n",
       "0            NaN  keratin 18 pseudogene 55 [Source:HGNC Symbol;A...   \n",
       "1       7.882377                                   3.33113425699397   \n",
       "\n",
       "                                Unnamed: 17733  \\\n",
       "0  hsa-mir-5195 [Source:miRBase;Acc:MI0018174]   \n",
       "1                             2.85253711034634   \n",
       "\n",
       "                                            POLRMTP1  \\\n",
       "0  polymerase (RNA) mitochondrial (DNA directed) ...   \n",
       "1                                   3.13069614031282   \n",
       "\n",
       "                                              UBL5P2  \\\n",
       "0  ubiquitin-like 5 pseudogene 2 [Source:HGNC Sym...   \n",
       "1                                   9.98661583035146   \n",
       "\n",
       "                                            TBC1D3P5 Unnamed: 17737  \n",
       "0  TBC1 domain family, member 3 pseudogene 5 [Sou...            NaN  \n",
       "1                                   3.07372352861326       7.284733  \n",
       "\n",
       "[2 rows x 17738 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr = pd.read_table('Cell_line_COSMIC_ID_gene_expression_transposed_clean.tsv', header=None)\n",
    "header = pd.read_table('Cell_line_RMA_proc_basalExp_transposed.tsv', nrows = 2)\n",
    "\n",
    "more_info = header\n",
    "more_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "expr.iloc[:,0] = expr.iloc[:,0].astype(int)\n",
    "expr.columns = header.columns\n",
    "expr = expr.rename(index=str, columns={\"GENE_SYMBOLS\": \"COSMIC_ID\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def hack(x):\n",
    "        return train_target.corr(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA & Linear Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# There are 265 different drugs - goal is to form 265 different models\n",
    "\n",
    "# Format the data frame to aggregate the gene expression info (left)\n",
    "# with the drug response info (right), where applicable \n",
    "\n",
    "# Note many NaNs throughout because not all drugs tested on all cell lines\n",
    "\n",
    "rmse_all = []\n",
    "\n",
    "for drug in dose_response['DRUG_ID'].unique():\n",
    "    subset = dose_response[ dose_response['DRUG_ID']== drug ].loc[:,['COSMIC_ID','LN_IC50']] # 'AUC', 'RMSE' \n",
    "    expre = expr.merge(subset, on='COSMIC_ID', how='left', sort=False)\n",
    "\n",
    "    # remove the id's \n",
    "    expre = expre.iloc[:,1:]\n",
    "    expr_data = expre.copy() \n",
    "    expr_target = expre.loc[:,'LN_IC50']\n",
    "    del expr_data['LN_IC50']\n",
    "\n",
    "    # remove na's\n",
    "    mask = np.array(~expr_target.isnull())\n",
    "    expr_target = expr_target[mask]\n",
    "    expr_data = pd.DataFrame(expr_data)[mask]\n",
    "    \n",
    "    # CV k= 5\n",
    "    rmse_drug = []\n",
    "    \n",
    "    for i in np.arange(0,5) :\n",
    "        \n",
    "        train_data, test_data, train_target, test_target = train_test_split( expr_data, expr_target, test_size=0.2)\n",
    "        # rownames are cosmic ids\n",
    "        # colnames are gene expression\n",
    "        \n",
    "        #auc and rmse filters\n",
    "        #auc_limit, rmse_limit = train_data[['AUC', 'RMSE']].quantile(0.5)\n",
    "        #filter_mask =  (train_data['AUC'] >= auc_limit)  & (train_data['RMSE'] <= rmse_limit) \n",
    "        #train_data = train_data[filter_mask]\n",
    "        #train_target = train_target[ filter_mask ]\n",
    "        \n",
    "        #remove auc, rmse after filtering\n",
    "        #del train_data['AUC']\n",
    "        #del train_data['RMSE']\n",
    "        #del test_data['AUC']\n",
    "        #del test_data['RMSE']\n",
    "        \n",
    "        #filter by correlation\n",
    "        #corr = train_data.apply(hack)\n",
    "        #top_corr = corr.quantile(0.8)\n",
    "        #low_corr = corr.quantile(0.2)\n",
    "        #Bool = (corr >= top_corr) | (corr <= low_corr)\n",
    "        #train_data = train_data[train_data.columns[Bool]]\n",
    "        #test_data = test_data[test_data.columns[Bool]]\n",
    "\n",
    "        if train_data.shape[1] ==0 or test_data.shape[1] ==0 :\n",
    "            rmse_drug.append(np.nan)\n",
    "            \n",
    "        else:\n",
    "\n",
    "            # scale the data\n",
    "            scaled_train = preprocessing.scale(train_data)\n",
    "            scaled_test = preprocessing.scale(test_data)\n",
    "            pca = PCA(.90)  \n",
    "\n",
    "            # fit the pca dim reduction on training data\n",
    "            train_img = pca.fit_transform(scaled_train)\n",
    "            \n",
    "            # apply it to testing data\n",
    "            test_img = pca.transform(scaled_test)\n",
    "\n",
    "            train_target = np.ravel(train_target)\n",
    "\n",
    "            linRegr = LinearRegression()\n",
    "            linRegr.fit(train_img, train_target)\n",
    "\n",
    "            pred = linRegr.predict(test_img)\n",
    "\n",
    "            mask = np.array(~test_target.isnull())\n",
    "            test_targ = test_target[mask]\n",
    "            pred = pd.DataFrame(pred)[mask]\n",
    "\n",
    "            if test_targ.shape[0] ==0 or pred.shape[0] ==0 : \n",
    "                rmse_drug.append(np.nan)\n",
    "\n",
    "            else:        \n",
    "                rmse_drug.append( np.sqrt(mean_squared_error(test_targ, pred) ))\n",
    "    \n",
    "    rmse_all.append(np.mean(rmse_drug)) \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model Comparisons (RMSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1262    0.446368\n",
       "266     0.458985\n",
       "150     0.472007\n",
       "1264    0.524581\n",
       "91      0.548536\n",
       "341     0.552194\n",
       "1502    0.555846\n",
       "205     0.641457\n",
       "1018    0.647145\n",
       "193     0.649277\n",
       "dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# best RMSE by drug\n",
    "rmse_all = pd.Series(rmse_all)\n",
    "rmse_all.index = dose_response['DRUG_ID'].unique()[:]\n",
    "rmse_all.sort_values().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3       2.060696\n",
       "299     2.071485\n",
       "302     2.077077\n",
       "344     2.083248\n",
       "51      2.157260\n",
       "190     2.214239\n",
       "346     2.237624\n",
       "268     2.247033\n",
       "1248    2.513362\n",
       "135     2.550817\n",
       "dtype: float64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# worst RMSE by drug\n",
    "rmse_all.sort_values().tail(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Random Forests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "accuracy_all = []\n",
    "f1_all = []\n",
    "auc_all = []\n",
    "\n",
    "for drug in dose_response['DRUG_ID'].unique():\n",
    "    subset = dose_response[ dose_response['DRUG_ID']== drug ].loc[:,['COSMIC_ID','LN_IC50']] #, 'AUC', 'RMSE'\n",
    "    expre = expr.merge(subset, on='COSMIC_ID', how='left', sort=False)\n",
    "\n",
    "    # remove the id's \n",
    "    expre = expre.iloc[:,1:]\n",
    "    expr_data = expre.copy() \n",
    "    expr_target = expre.loc[:,'LN_IC50'] \n",
    "    del expr_data['LN_IC50']\n",
    "\n",
    "    # remove na's\n",
    "    mask = np.array(~expr_target.isnull())\n",
    "    expr_target = expr_target[mask]\n",
    "    expr_data = pd.DataFrame(expr_data)[mask]\n",
    "    \n",
    "    # CV k= 5 - cross validation is not necessary for random forests because of the averaging throughout!\n",
    "    # Remove cv loops in future\n",
    "    accuracy_drug = []\n",
    "    f1_drug = []\n",
    "    auc_drug = []\n",
    "\n",
    "    for i in np.arange(0,5) :\n",
    "        train_data, test_data, train_target, test_target = train_test_split( expr_data, expr_target, test_size=0.2)\n",
    "        \n",
    "        #remove null values\n",
    "        mask = np.array(~train_target.isnull())\n",
    "        train_target = train_target[mask]\n",
    "        train_data = train_data[mask]\n",
    "        \n",
    "        #set the SIR categories\n",
    "        sir_one_third = train_target.quantile(1/3)\n",
    "        sir_two_third = train_target.quantile(2/3)\n",
    "        \n",
    "        #assign SIR classification\n",
    "        train_target = pd.DataFrame(train_target > sir_two_third)*2 + pd.DataFrame(train_target < sir_one_third)*1 \n",
    "        train_target = train_target.astype(str)\n",
    "        \n",
    "        #auc and rmse filters\n",
    "        #auc_limit, rmse_limit = train_data[['AUC', 'RMSE']].quantile(0.5)\n",
    "        #filter_mask =  (train_data['AUC'] >= auc_limit)  & (train_data['RMSE'] <= rmse_limit) \n",
    "        #train_data = train_data[filter_mask]\n",
    "        #train_target = train_target[ filter_mask ]\n",
    "        \n",
    "        #remove auc, rmse after filtering\n",
    "        #del train_data['AUC']\n",
    "        #del train_data['RMSE']\n",
    "        #del test_data['AUC']\n",
    "        #del test_data['RMSE']\n",
    "        \n",
    "        #filter by correlation\n",
    "        #corr = train_data.apply(hack)\n",
    "        #top_corr = corr.quantile(0.8)\n",
    "        #low_corr = corr.quantile(0.2)\n",
    "        #Bool = (corr >= top_corr) | (corr <= low_corr)\n",
    "        #train_data = train_data[train_data.columns[Bool]]\n",
    "        #test_data = test_data[test_data.columns[Bool]]\n",
    "\n",
    "        if train_data.shape[1] ==0 or test_data.shape[1] ==0 :\n",
    "            accuracy_drug.append(np.nan)\n",
    "            f1_drug.append(np.nan)\n",
    "            \n",
    "        else:\n",
    "\n",
    "            #PCA for dim reduction - fit on training data only\n",
    "            #scaled_train = preprocessing.scale(train_data)\n",
    "            #scaled_test = preprocessing.scale(test_data)\n",
    "            #pca = PCA(.90)  \n",
    "            #train_img = pca.fit_transform(scaled_train)\n",
    "            #test_img = pca.transform(scaled_test)\n",
    "\n",
    "            train_target = np.ravel(train_target)\n",
    "\n",
    "            rf = RandomForestClassifier(n_estimators=30, oob_score=True, random_state=0)\n",
    "            rf.fit(train_data, train_target)\n",
    "            \n",
    "            mask = np.array(~test_target.isnull())\n",
    "            test_target = test_target[mask]\n",
    "            test_data = test_data[mask]\n",
    "            \n",
    "            #assign SIR classification\n",
    "            test_target = pd.DataFrame(test_target > sir_two_third)*2 + pd.DataFrame(test_target < sir_one_third)*1 \n",
    "            test_target = test_target.astype(str)\n",
    "\n",
    "            pred = rf.predict(test_data)\n",
    "            \n",
    "            if test_target.shape[0] ==0 or pred.shape[0] ==0 : \n",
    "                accuracy_drug.append(np.nan)\n",
    "                f1_drug.append(np.nan)\n",
    "\n",
    "            else:        \n",
    "                            \n",
    "                test_target_bin = label_binarize(test_target, classes=['0','1','2'])\n",
    "                pred_bin = label_binarize(pred, classes=['0','1','2'])\n",
    "\n",
    "                fpr = dict()\n",
    "                tpr = dict()\n",
    "                roc_auc = [0,0,0]\n",
    "                for i in range(3):\n",
    "                    fpr[i], tpr[i], _ = roc_curve(test_target_bin[:, i], pred_bin[:, i])\n",
    "                    roc_auc[i] = np.round( auc(fpr[i], tpr[i]), 2 )\n",
    "                \n",
    "                accuracy_drug.append( accuracy_score(test_target, pred) )\n",
    "                f1_drug.append( f1_score(test_target, pred, average='macro') )\n",
    "                auc_drug.append(roc_auc)\n",
    "\n",
    "    accuracy_all.append(np.mean(accuracy_drug)) \n",
    "    f1_all.append(np.mean(f1_drug)) \n",
    "    auc_all.append(np.mean(auc_drug, axis=0)) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Model Comparisons (Accuracy, F1, AUC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accuracy</th>\n",
       "      <th>f1</th>\n",
       "      <th>auc[0,1,2]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>0.596</td>\n",
       "      <td>0.590</td>\n",
       "      <td>[0.628, 0.6799999999999999, 0.78]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1372</th>\n",
       "      <td>0.593</td>\n",
       "      <td>0.589</td>\n",
       "      <td>[0.612, 0.682, 0.79]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>0.558</td>\n",
       "      <td>0.558</td>\n",
       "      <td>[0.592, 0.766, 0.6439999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>0.549</td>\n",
       "      <td>0.550</td>\n",
       "      <td>[0.588, 0.682, 0.7220000000000001]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1014</th>\n",
       "      <td>0.549</td>\n",
       "      <td>0.547</td>\n",
       "      <td>[0.582, 0.6860000000000002, 0.724]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>0.539</td>\n",
       "      <td>0.536</td>\n",
       "      <td>[0.612, 0.726, 0.622]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1047</th>\n",
       "      <td>0.536</td>\n",
       "      <td>0.525</td>\n",
       "      <td>[0.558, 0.74, 0.6479999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>0.531</td>\n",
       "      <td>0.529</td>\n",
       "      <td>[0.584, 0.6940000000000001, 0.668]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>0.528</td>\n",
       "      <td>0.532</td>\n",
       "      <td>[0.5599999999999999, 0.7619999999999999, 0.618...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>0.521</td>\n",
       "      <td>0.526</td>\n",
       "      <td>[0.5599999999999999, 0.75, 0.608]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accuracy     f1                                         auc[0,1,2]\n",
       "1526     0.596  0.590                  [0.628, 0.6799999999999999, 0.78]\n",
       "1372     0.593  0.589                               [0.612, 0.682, 0.79]\n",
       "211      0.558  0.558                 [0.592, 0.766, 0.6439999999999999]\n",
       "1498     0.549  0.550                 [0.588, 0.682, 0.7220000000000001]\n",
       "1014     0.549  0.547                 [0.582, 0.6860000000000002, 0.724]\n",
       "1377     0.539  0.536                              [0.612, 0.726, 0.622]\n",
       "1047     0.536  0.525                  [0.558, 0.74, 0.6479999999999999]\n",
       "252      0.531  0.529                 [0.584, 0.6940000000000001, 0.668]\n",
       "253      0.528  0.532  [0.5599999999999999, 0.7619999999999999, 0.618...\n",
       "290      0.521  0.526                  [0.5599999999999999, 0.75, 0.608]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# best accuracy by drug\n",
    "\n",
    "accuracy = pd.Series(accuracy_all)\n",
    "f1 = pd.Series(f1_all)\n",
    "auc = pd.Series(auc_all)\n",
    "\n",
    "new = pd.DataFrame({'accuracy': accuracy.round(3), 'f1': f1.round(3), 'auc[0,1,2]': auc })\n",
    "new.index = dose_response['DRUG_ID'].unique()[:]\n",
    "new.sort_values(by='accuracy', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accuracy</th>\n",
       "      <th>f1</th>\n",
       "      <th>auc[0,1,2]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>0.371</td>\n",
       "      <td>0.370</td>\n",
       "      <td>[0.508, 0.522, 0.5599999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>0.370</td>\n",
       "      <td>0.366</td>\n",
       "      <td>[0.534, 0.538, 0.5199999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>0.370</td>\n",
       "      <td>0.371</td>\n",
       "      <td>[0.512, 0.5199999999999999, 0.554]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>0.369</td>\n",
       "      <td>0.367</td>\n",
       "      <td>[0.51, 0.52, 0.5519999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>0.365</td>\n",
       "      <td>0.359</td>\n",
       "      <td>[0.5079999999999999, 0.544, 0.522]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>0.364</td>\n",
       "      <td>0.361</td>\n",
       "      <td>[0.48600000000000004, 0.508, 0.574]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053</th>\n",
       "      <td>0.359</td>\n",
       "      <td>0.356</td>\n",
       "      <td>[0.5, 0.5599999999999999, 0.5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>0.358</td>\n",
       "      <td>0.357</td>\n",
       "      <td>[0.496, 0.51, 0.5519999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>0.350</td>\n",
       "      <td>0.348</td>\n",
       "      <td>[0.4880000000000001, 0.506, 0.5399999999999999]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>0.344</td>\n",
       "      <td>0.339</td>\n",
       "      <td>[0.488, 0.55, 0.48999999999999994]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      accuracy     f1                                       auc[0,1,2]\n",
       "204      0.371  0.370               [0.508, 0.522, 0.5599999999999999]\n",
       "63       0.370  0.366               [0.534, 0.538, 0.5199999999999999]\n",
       "197      0.370  0.371               [0.512, 0.5199999999999999, 0.554]\n",
       "186      0.369  0.367                 [0.51, 0.52, 0.5519999999999999]\n",
       "110      0.365  0.359               [0.5079999999999999, 0.544, 0.522]\n",
       "157      0.364  0.361              [0.48600000000000004, 0.508, 0.574]\n",
       "1053     0.359  0.356                   [0.5, 0.5599999999999999, 0.5]\n",
       "207      0.358  0.357                [0.496, 0.51, 0.5519999999999999]\n",
       "202      0.350  0.348  [0.4880000000000001, 0.506, 0.5399999999999999]\n",
       "35       0.344  0.339               [0.488, 0.55, 0.48999999999999994]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# worst accuracy by drug\n",
    "\n",
    "new.sort_values(by='accuracy', ascending=False).tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
